source/extensions/transport_sockets/tls/ssl_socket.h

```diff
@@ -77,9 +79,9 @@ class SslSocketInfo : public Envoy::Ssl::ConnectionInfo {
   std::string ciphersuiteString() const override;
   const std::string& tlsVersion() const override;
   absl::optional<std::string> x509Extension(absl::string_view extension_name) const override;
-  SSL* ssl() const { return ssl_.get(); }
+  SSL* ssl() const { return ssl_; }
 
-  bssl::UniquePtr<SSL> ssl_;
+  SSL* ssl_;
```

Comment:
This will cause ssl_ to be dangling in logging phase after the connection is terminated. We used to own the SSL object here to make sure it outlives StreamInfo, which is holding a shared pointer of SslSocketInfo.

Reply:
Great catch. Rather than try and transfer bssl::UniquePtr<SSL> ownership from Handshaker to SslSocketInfo at dtor time, I've chosen to give both SslSocket and SslSocketInfo a shared_ptr to the Handshaker, so that the Handshaker lives until _both_ the SslSocket and the SslSocketInfo are destroyed. I think this solved the problem, PTAL.